app/Http/Controllers/Api/VolumeController.php

```diff
@@ -209,4 +209,36 @@ public function clone(CloneVolume $request)
             ->with('messageType', 'success');
 
     }
+
+    /**
+     * Return volume's annotation timestamps of newest annotations
+     *
+     * @param int $id
+     * @return object
+     * @api {get} volumes{/id}/files/annotation-timestamps Get the newest annotation timestamps of the volume
+     * @apiGroup Volumes
+     * @apiPermission projectMember
+     *
+     * @apiParam {Number} id The volume ID.
+     *
+     * @apiSuccessExample {json} Success response:
+     * {
+     *  1: "2024-08-23T07:06:31.000000Z",
+     *  2: "2024-08-23T06:40:29.000000Z",
+     *  3: "2024-08-23T06:40:35.000000Z",
+     * }
+     *
+     */
+    public function getAnnotationTimestamps($id)
+    {
+        $volume = Volume::findOrFail($id);
+        $this->authorize('access', $volume);
+
+        $timestamps = $volume->files
+            ->flatMap(fn ($file) => $file->annotations)
+            ->groupBy('file_id')
+            ->map(fn ($annotations) => $annotations->max('created_at'));
```

Comment:
It may be even better to just return the sorted file IDs from this endpoint. Then you don't need to do the sorting in the JS and transfer less data.

Also I challenge you to find a single database query that sorts the files by the newest annotation timestamp and also includes files without annotations (at the end). Hints: `groupBy` (again) and `leftJoin` :wink: This will make the API endpoint much more efficient, as it does not have to process possibly hundreds of thousands of annotations.

Reply:
Can you explain, why this approach would not process a large amount of annotations?

Reply:
There are three issues:

1. `$volume->files`: This loads all files of the volume into a collection. Volumes can have many thousands of files so this may already consume a lot of memory. Often building a query with `$volume->files()` is better but this wouldn't work here, too.

2. `->flatMap(fn ($file) => $file->annotations)`: This replaces the files collection with a collection of annotations. *All* annotations in the volume. This can obviously be huge as well.

3. `->map(fn ($annotations) => $annotations->max('created_at'))`: This executes a database query for each file. With thousands of files, you have thousands of queries.

Let the database do all the work, it will always be faster and more memory efficient! What you do with 2*n*+1 queries (with *n* the number of files) can also be done with 1 query.